[TT-12266] Add Rate Limit Smoothing Docs (5.4.0) (#4771)
TykTechnologies/tyk-docs: tyk-docs/content/getting-started/key-concepts/rate-limiting.md

---
title: Rate Limiting in Tyk
tags:

>     - Rate Limit

Author: Whitespace

- do you want to limit the number of requests a specific user can make to **all APIs** they have access to? **You’ll want to use a key-level global rate limit!**
- do you want to limit the number of requests a specific user can make to **specific APIs** they have access to? **You’ll want to use a key-level per-API rate limit.**

### Applying multiple rate limits

When multiple rate limits are configured, they are assessed in this order (if applied):

1. API-level global rate limit
2. Key-level global rate limit
3. Key-level per-API rate limit

## Rate limiting algorithms

Different rate limiting algorithms are employed to cater to varying requirements, use cases and gateway deployments. A one-size-fits-all approach may not be suitable, as APIs can have diverse traffic patterns, resource constraints, and service level objectives. Some algorithms are more suited to protecting the upstream service from overload whilst others are suitable for per-client limiting to manage and control fair access to a shared resource.

Tyk offers the following rate limiting algorithms:

> 1. [Distributed Rate Limiter]({{< ref "#Distributed-Rate-Limiter" >}}): recommended for most use cases, implements the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket)

Reviewer: ```suggestion
1. [Distributed Rate Limiter]({{< ref "#distributed-rate-limiter" >}}): recommended for most use cases, implements the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket)
```

> 2. [Redis Rate Limiter]({{< ref "#Redis-Rate-Limiter" >}}): implements the [sliding window log algorithm](https://developer.redis.com/develop/dotnet/aspnetcore/rate-limiting/sliding-window)

Reviewer: ```suggestion
2. [Redis Rate Limiter]({{< ref "#redis-rate-limiter" >}}): implements the [sliding window log algorithm](https://developer.redis.com/develop/dotnet/aspnetcore/rate-limiting/sliding-window)
```

- configure [Rate Limit Smoothing]({{< ref "#rate-limit-smoothing" >}}) to gradually adjust limits.

##### Rate Limit Smoothing

Rate Limit Smoothing is a mechanism to dynamically adjust the request
rate limits based on the current traffic patterns. It helps in managing
request spikes by gradually increasing or decreasing the rate limit
instead of making abrupt changes or blocking requests excessively.

When Redis Rate Limiter is in use, rate limit smoothing may be configured
on a key level, policy level, or per-API with the following options:

- `enabled` (boolean) to enable or disable rate limit smoothing
- `threshold` after which to apply smoothing (minimum rate for window)
- `trigger` configures at which fraction of a step a smoothing event is emitted
- `step` is the value by which the rate allowance will get adjusted
- `delay` is the amount of seconds between smoothing updates

Rate Limit Smoothing is configured using:

- API Keys, under `.smoothing.*`
- Policies, under `.smoothing.*`
- Per-API, via Keys and Policies under `.access_rights[<api_id>].limit.smoothing.*`

An example configuration would be as follows:

> ```json

Reviewer 2: ```suggestion
```yaml
```

"smoothing": {
  "enabled": true,
  "threshold": 5,
  "trigger": 0.5,
  "step": 5,
  "delay": 30
}
```

This is used to compute a request allowance. The request allowance will
be smoothed between `threshold`, and the defined rate limits (maximum).
The request allowance will be updated internally every `delay` seconds.

Events are emitted based on the following calculations:

- When the request rate rises above `allowance - (step * trigger)`,

>   a RateLimitSmoothingUp event is emitted and allowance increases by `step`.

Reviewer 2: ```suggestion
  a *RateLimitSmoothingUp* event is emitted and allowance increases by `step`.
```

Reviewer: This is no longer relevant following re-write of the page. Closing comment.

- When the request rate falls below `allowance - (step + step * trigger)`,

>   a RateLimitSmoothingDown event is emitted and allowance decreases by `step`.

Reviewer 2: ```suggestion
  a *RateLimitSmoothingDown* event is emitted and allowance decreases by `step`.
```

Reviewer: This is no longer relevant following re-write of the page. Closing comment.

> To see how to configure the value on keys or policies:

Reviewer 2: ```suggestion
Please consult the following guides for an explanation of how to configure the value on keys or policies:
```
